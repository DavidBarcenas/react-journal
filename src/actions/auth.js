import { types } from "../types/types";

export const startLoginEmailPwd = ( email, pwd ) => {
  return ( dispatch ) => {
    setTimeout(() => {
      dispatch(login(123, 'Daveepro'))
    }, 3500);    
  }
}

export const login = ( uid, displayName ) => ({
  type: types.login,
  payload: {
    uid, 
    displayName
  }
})