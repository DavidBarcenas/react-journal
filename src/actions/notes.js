import { db } from "../firebase/firebase-provider";
import { types } from "../types/types";
import { loadNotes } from "../helpers/loadNotes";
import Swal from "sweetalert2";
import { fileUpload } from "../helpers/fileUpload";

export const startNewNote = () => {
  return async ( dispatch, getState ) => {
    const {auth: { uid }} = getState();
    const newNote = {
      title: '',
      body: '',
      date: new Date().getTime()
    }
    const doc = await db.collection(`${ uid }/journal/notes`).add( newNote )
    dispatch(activeNote( doc.id, newNote ))
  }
}

export const activeNote = ( id, note ) => ({
  type: types.noteActive,
  payload: {
    id, 
    ...note
  }
})

export const setNotes = ( notes ) => ({
  type: types.noteLoad,
  payload: notes
})

export const startLoadingNotes = ( uid ) => {
  return async ( dispatch ) => {
    const notes = await loadNotes( uid )
    dispatch(setNotes( notes ))
  }
}

export const startSaveNote = ( note ) => {
  return async ( dispatch, getState ) => {
    const {auth: { uid }} = getState();

    if( !note.url ) { delete note.url }

    const noteToFS = { ...note }
    delete noteToFS.id;

    await db.doc(`${ uid }/journal/notes/${ note.id }`).update(noteToFS)
    dispatch(refreshNote( note.id, noteToFS ))
    Swal.fire('Saved', note.title, 'success')
  }
}

export const refreshNote = ( id, note ) => ({
  type: types.noteUpdated,
  payload: {
    id, note: { id, ...note }
  }
})

export const startUploading = ( file ) => {
  return async ( dispatch, getState ) => {
    const { active: activeNote } = getState().notes;
    const fileUrl = await fileUpload( file )

    console.log('fileUrl', fileUrl)
  }
}