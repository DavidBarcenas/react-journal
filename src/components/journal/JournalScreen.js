import React from 'react'
import { Sidebar } from './Sidebar'

export const JournalScreen = () => {
  return (
    <div className="joutnal__wrapper">
      <Sidebar />
      <main><h1>Main content</h1></main>
    </div>
  )
}
