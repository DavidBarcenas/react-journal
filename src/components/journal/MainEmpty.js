import React from 'react'

export const MainEmpty = () => {
  return (
    <div className="main__empty">
      <p>Select something</p>
      <p>or create an entry</p>
      <i className="far fa-star fa-4x"></i>
    </div>
  )
}
