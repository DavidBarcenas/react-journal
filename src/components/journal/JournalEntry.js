import React from 'react'
import { useDispatch } from 'react-redux'
import { format } from 'date-fns'
import { activeNote } from '../../actions/notes'

export const JournalEntry = ({ body, date, title, id, url }) => {

  const dispatch = useDispatch()

  const handleEntryClick = () => {
    dispatch(activeNote( id, { body, date, title, id, url } ))
  }
  
  return (
    <div className="animate__animated animate__fadeInUp animate__faster journal__entry" onClick={ handleEntryClick }>
      { url &&
        <div className="journal__entry-picture" style={{backgroundImage: `url(${ url })`}}> </div>
      }
        

      <div className="journal__entry-body">
        <p className="journal__entry-title">{ title }</p>
        <p className="journal__entry-content">{ body }</p>
      </div>

      <div className="journal__entry-date">
        <span className="journal__entry-day">{ format(new Date( date ), "EEEE") }</span>
        <span className="journal__entry-num">{ format(new Date( date ), "io") }</span>
      </div>
    </div>
  )
}
