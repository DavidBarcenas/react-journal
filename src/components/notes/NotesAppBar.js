import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { startSaveNote } from '../../actions/notes'

export const NotesAppBar = () => {

  const dispatch = useDispatch()
  const { active } = useSelector(state => state.notes)

  const handleSave = () => {
    dispatch(startSaveNote( active ))
  }

  return (
    <div className="notes__appbar">
      <span className="notes__appbar-date"> 28 de Agosto de 2020 </span>
      <div>
        <button className="btn"> pic </button>
        <button className="btn" onClick={ handleSave }> Save </button>
      </div>
    </div>
  )
}
