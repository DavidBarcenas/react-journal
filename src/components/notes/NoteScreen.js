import React, { useEffect, useRef } from 'react'
import { NotesAppBar } from './NotesAppBar'
import { useDispatch, useSelector } from 'react-redux'
import { useForm } from '../../hooks/useForm'
import { activeNote } from '../../actions/notes'

export const NoteScreen = () => {

  const dispatch = useDispatch()
  const { active:note } = useSelector( state => state.notes)
  const activeId = useRef( note.id )
  const [ formValues, handleInputChange, reset ] = useForm(note)

  useEffect(() => {
    if( note.id !== activeId.current ) {
      reset(note)
      activeId.current = note.id
    }
  }, [ note, reset ])

  useEffect(() => {
    dispatch(activeNote(formValues.id, {...formValues}))
  }, [formValues, dispatch])

  return (
    <div className="notes__main">
      <NotesAppBar />

      <div className="notes__content">
        <form>
          <input type="text" placeholder="Some awesome title" name="title" className="notes__input" value={ formValues.title } onChange={ handleInputChange } />
          <textarea name="body" cols="5" rows="10" placeholder="What happend today" className="notes__textarea" value={ formValues.body } onChange={ handleInputChange } ></textarea>
        </form>

        <div className="notes__image">
          { note.url &&
            <img src={ note.url } alt={ note.title } />
          }
        </div>
      </div>
    </div>
  )
}
