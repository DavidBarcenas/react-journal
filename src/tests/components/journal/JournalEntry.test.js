import '@testing-library/jest-dom';
import React from 'react'
import configureStore from 'redux-mock-store'
import { JournalEntry } from '../../../components/journal/JournalEntry';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { activeNote } from '../../../actions/notes';

const note = {
  id: 10,
  date: 0,
  title: 'Hello',
  body: 'World',
  url: 'https://test/image.png'
}

const mockStore = configureStore()
let store = mockStore({})
store.dispatch = jest.fn()

const wrapper = mount(
  <Provider store={ store }>
      <JournalEntry { ...note } />
  </Provider>
)
describe('<JurnalEntry /> testing', () => {
  test('should render correctly', () => {
    expect(wrapper).toMatchSnapshot()
  })
  
  test('should trigger note action', () => {
    wrapper.find('.journal__entry').prop('onClick')()
    expect( store.dispatch ).toHaveBeenCalledWith(
      activeNote( note.id, { ...note } )
    )
  })
  
})
