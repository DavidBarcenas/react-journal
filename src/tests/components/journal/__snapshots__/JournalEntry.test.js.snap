// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<JurnalEntry /> testing should render correctly 1`] = `
<div
  className="animate__animated animate__fadeInUp animate__faster journal__entry"
  onClick={[Function]}
>
  <div
    className="journal__entry-picture"
    style={
      Object {
        "backgroundImage": "url(https://test/image.png)",
      }
    }
  >
     
  </div>
  <div
    className="journal__entry-body"
  >
    <p
      className="journal__entry-title"
    >
      Hello
    </p>
    <p
      className="journal__entry-content"
    >
      World
    </p>
  </div>
  <div
    className="journal__entry-date"
  >
    <span
      className="journal__entry-day"
    >
      Wednesday
    </span>
    <span
      className="journal__entry-num"
    >
      3rd
    </span>
  </div>
</div>
`;
